R/PipeOpVtreat.R

```diff
@@ -0,0 +1,217 @@
+        if (length(task$class_names) > 2L) {
+          treatment = vtreat::MultinomialOutcomeTreatment
+          params = vtreat::multinomial_parameters(self$param_set$values$params)
+        } else {
+          treatment = vtreat::BinomialOutcomeTreatment
+          params = vtreat::classification_parameters(self$param_set$values$params)
+        }
+      }
+
+      if (length(self$param_set$values$cols_to_copy)) {
+        checkmate::assert_subset(unlist(self$param_set$values$cols_to_copy), choices = var_list, empty.ok = TRUE)
+      }
+
+      if (length(self$param_set$values$imputation_map)) {
+        checkmate::assert_subset(names(self$param_set$values$imputation_map), choices = var_list, empty.ok = TRUE)
+      }
+
+      transform_design = mlr3misc::invoke(treatment,
+        var_list = var_list,
+        outcome_name = task$target_names,
+        cols_to_copy = self$param_set$values$cols_to_copy,
+        params = params,
+        imputation_map = self$param_set$values$imputation_map)
+
+      # the following exception handling is necessary because vtreat sometimes fails with "no usable vars" if the data is already "clean" enough
+      vtreat_res = try(
+        mlr3misc::invoke(vtreat::fit_prepare,
+          vps = transform_design,
+          dframe = task$data(),
+          weights = task$weights$weight),
+        silent = TRUE
+      )
+
+      if (class(vtreat_res) == "try-error") {
+        self$state = NULL
+        return(task)  # early exit
+      }
+
+      self$state$treatment_plan = vtreat_res$treatments
+
+      d_prepared = data.table::setDT(vtreat_res$cross_frame)
+
+      task$cbind(d_prepared)
+
+      feature_subset = self$state$treatment_plan$get_feature_names()  # subset to vtreat features
+
+      if (self$param_set$values$recommended) {
+        score_frame = mlr3misc::invoke(vtreat::get_score_frame, vps = self$state$treatment_plan)
+        feature_subset[feature_subset %in% score_frame$varName[score_frame$recommended]]  # subset to only recommended
+      }
+
+      feature_subset = c(feature_subset, unlist(self$state$treatment_plan$settings$cols_to_copy))  # respect cols_to_copy
+
+      task$select(feature_subset)
+
+      task
+    },
+
+    .predict_task = function(task) {
+      if (is.null(self$state$treatment_plan) || (length(task$feature_names) == 0L)) return(task)  # early exit
+
+      # in the following we suppress warnings and catch errors
+      # because if we predict on the same task that was used for training vtreat will tell us:
+      #"possibly called transform() on same data frame as fit(), this can lead to over-fit. To avoid this, please use fit_transform()."
+      # but this is fine, because we are in the predict step and all models are trained already
+      # also, in rare case vtreat will have dropped all features (if recommended = TRUE) and now will fail with "no usable vars"
+      d_prepared = try(
+        suppressWarnings(mlr3misc::invoke(vtreat::prepare,
```

Comment:
maybe check if suppressWarnings can be avoided without turning ugly, otherwise leave as is